src/TestFramework/TestFramework/Assertions/Assert.HasCount.cs

```diff
@@ -0,0 +1,120 @@
+        public void AppendFormatted(ReadOnlySpan<char> value, int alignment = 0, string? format = null)
+            => AppendFormatted(value.ToString(), alignment, format);
+#pragma warning restore RS0027 // API with optional parameter(s) should have the most parameters amongst its public overloads
+#endif
+
+        // NOTE: All the overloads involving format and/or alignment are not super efficient.
+        // This code path is only for when an assert is failing, so that's not the common scenario
+        // and should be okay if not very optimized.
+        // A more efficient implementation that can be used for .NET 6 and later is to delegate the work to
+        // the BCL's StringBuilder.AppendInterpolatedStringHandler
+        public void AppendFormatted<T>(T value, string? format)
+            => _builder!.AppendFormat(null, $"{{0:{format}}}", value);
+
+        public void AppendFormatted<T>(T value, int alignment)
+            => _builder!.AppendFormat(null, $"{{0,{alignment}}}", value);
+
+        public void AppendFormatted<T>(T value, int alignment, string? format)
+            => _builder!.AppendFormat(null, $"{{0,{alignment}:{format}}}", value);
+
+        public void AppendFormatted(string? value)
+            => _builder!.Append(value);
+
+#pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
+#pragma warning disable RS0027 // API with optional parameter(s) should have the most parameters amongst its public overloads
+        public void AppendFormatted(string? value, int alignment = 0, string? format = null)
+            => _builder!.AppendFormat(null, $"{{0,{alignment}:{format}}}", value);
+
+        public void AppendFormatted(object? value, int alignment = 0, string? format = null)
+            => _builder!.AppendFormat(null, $"{{0,{alignment}:{format}}}", value);
+#pragma warning restore RS0026 // Do not add multiple public overloads with optional parameters
+#pragma warning restore RS0027 // API with optional parameter(s) should have the most parameters amongst its public overloads
+    }
+
+    public static void HasCount<T>(int expectedCount, IEnumerable<T> collection)
+        => HasCount(expectedCount, collection, string.Empty, null);
+
+    public static void HasCount<T>(int expectedCount, IEnumerable<T> collection, string? message)
+        => HasCount(expectedCount, collection, message, null);
+
+#pragma warning disable IDE0060 // Remove unused parameter
+    public static void HasCount<T>(int expectedCount, IEnumerable<T> collection, [InterpolatedStringHandlerArgument(nameof(expectedCount), nameof(collection))] ref AssertHasCountInterpolatedStringHandler<T> message)
+#pragma warning restore IDE0060 // Remove unused parameter
+        => message.ComputeAssertion();
+
+    public static void HasCount<T>(int expected, IEnumerable<T> collection, [StringSyntax(StringSyntaxAttribute.CompositeFormat)] string? message, params object?[]? parameters)
```

Comment:
Other overloads name the `expectedCount`, but this one names it `expected`

Reply:
As discussed offline, going with `expected` everywhere